tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,482 @@
+{
+    // Default behavior of the eBPF program is to block the connection.
+
+    // Send should fail as the connection is blocked.
+    sender_socket->send_message_to_remote_host(CLIENT_MESSAGE, destination, _globals.remote_port);
+    sender_socket->complete_async_send(1000, expected_result_t::failure);
+
+    // Receive should timeout as connection is blocked.
+    sender_socket->post_async_receive(true);
+    sender_socket->complete_async_receive(1000, true);
+
+    // Now update the policy map to allow the connection and test again.
+    connect_redirect_test(object, sender_socket, destination, destination, dual_stack);
+}
+
+void
+delete_sender_socket(_In_ sender_socket_t* sender_socket)
+{
+    if (_globals.protocol == IPPROTO_TCP) {
+        delete ((stream_sender_socket_t*)sender_socket);
+    } else {
+        delete ((datagram_sender_socket_t*)sender_socket);
+    }
+}
+
+void
+get_sender_socket(bool dual_stack, _Inout_ sender_socket_t** sender_socket)
+{
+    sender_socket_t* old_socket = *sender_socket;
+    sender_socket_t* new_socket = nullptr;
+    socket_family_t family = dual_stack
+                                 ? socket_family_t::Dual
+                                 : ((_globals.family == AF_INET) ? socket_family_t::IPv4 : socket_family_t::IPv6);
+    if (_globals.protocol == IPPROTO_TCP) {
+        new_socket = (sender_socket_t*)new stream_sender_socket_t(SOCK_STREAM, IPPROTO_TCP, 0, family);
+    } else {
+        new_socket = (sender_socket_t*)new datagram_sender_socket_t(SOCK_DGRAM, IPPROTO_UDP, 0, family);
+    }
+
+    *sender_socket = new_socket;
+    if (old_socket) {
+        delete_sender_socket(old_socket);
+    }
+}
+
+void
+authorize_test_wrapper(_In_ const struct bpf_object* object, bool dual_stack, _In_ sockaddr_storage& destination)
+{
+    sender_socket_t* sender_socket = nullptr;
+
+    get_sender_socket(dual_stack, &sender_socket);
+    authorize_test(object, sender_socket, destination, dual_stack);
+    delete_sender_socket(sender_socket);
+}
+
+void
+connect_redirect_test_wrapper(
+    _In_ const struct bpf_object* object,
+    _In_ sockaddr_storage& destination,
+    _In_ sockaddr_storage& proxy,
+    bool dual_stack)
+{
+    sender_socket_t* sender_socket = nullptr;
+    get_sender_socket(dual_stack, &sender_socket);
+    connect_redirect_test(object, sender_socket, destination, proxy, dual_stack);
+    delete_sender_socket(sender_socket);
+}
+
+void
+connect_redirect_tests_common(_In_ const struct bpf_object* object, bool dual_stack, _In_ test_addresses_t& addresses)
+{
+    // First cateogory is authorize tests.
+    const char* protocol_string = (_globals.protocol == IPPROTO_TCP) ? "TCP" : "UDP";
+    const char* family_string = (_globals.family == AF_INET) ? "IPv4" : "IPv6";
+    const char* dual_stack_string = dual_stack ? "Dual Stack" : "No Dual Stack";
+
+    // Loopback address.
+    printf("AUTH: Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
```

Comment:
Do these go to logs? Can you use some Catch macro instead?

Reply:
Yes these are visible in raw logs in CICD.